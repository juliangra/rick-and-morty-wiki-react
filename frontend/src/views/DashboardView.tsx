import { Outlet, useParams } from 'react-router'
import { Box, Center, Container, Grid, Pagination, Text, TextInput } from '@mantine/core'
import CharacterCard from 'src/components/CharacterCard'
import { useGetCharactersQuery } from 'src/graphql/generated/generated'
import { useDebouncedValue } from '@mantine/hooks'
import { useEffect, useState } from 'react'
import CustomLoading from 'src/components/common/CustomLoading'

const DashboardView = () => {
  const { id } = useParams()
  const [page, setPage] = useState(1)
  const [searchInput, setSearchInput] = useState('')
  const [debouncedSearchInput] = useDebouncedValue(searchInput, 250)

  const { data, loading, refetch } = useGetCharactersQuery({
    variables: { page }
  })

  useEffect(() => {
    refetch({ page: page, filter: { name: debouncedSearchInput } })
  }, [page, debouncedSearchInput])

  const characters = data?.characters?.results
  const numberOfPages = data?.characters?.info?.pages || 0

  if (loading) {
    return <CustomLoading />
  }

  return (
    <Box>
      {!id ? (
        <Container>
          <Text size="xl">Characters</Text>
          <TextInput
            label="Search for a character"
            value={searchInput}
            style={{ flex: 1 }}
            my={24}
            onChange={(event) => setSearchInput(event.currentTarget.value)}
            data-testid="search-input"
          />
          <Grid data-testid="characters-container" data-cy="characters-container">
            {characters?.map((character) => (
              <Grid.Col span={4} key={character?.id}>
                <CharacterCard character={character} />
              </Grid.Col>
            ))}
          </Grid>

          <Center style={{ marginTop: 24, marginBottom: 24 }}>
            <Pagination page={page} onChange={setPage} total={numberOfPages} size="xl" />
          </Center>
        </Container>
      ) : (
        <Outlet />
      )}
    </Box>
  )
}

export default DashboardView
