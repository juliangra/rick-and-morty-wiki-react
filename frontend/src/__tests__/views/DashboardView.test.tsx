import { screen } from '@testing-library/react'
import DashboardView from '../../views/DashboardView'
import { createWithApolloClient, renderWithApolloClient } from '../../utils/tests/render'

describe('DashboardView', () => {
  it('matches snapshot', () => {
    const component = createWithApolloClient(<DashboardView />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('initially loads with loading state', async () => {
    await renderWithApolloClient(<DashboardView />)
    expect(screen.getByTestId('loading')).toBeInTheDocument()
  })
})

export {}
