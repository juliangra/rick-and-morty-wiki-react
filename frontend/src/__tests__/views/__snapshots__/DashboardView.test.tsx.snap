// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardView matches snapshot 1`] = `
<div
  className="mantine-8od8ev"
  style={
    Object {
      "minHeight": "100vh",
    }
  }
>
  <svg
    className="mantine-1avyp1d"
    data-testid="loading"
    height="36px"
    role="presentation"
    stroke="#4c6ef5"
    viewBox="0 0 38 38"
    width="36px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="none"
      fillRule="evenodd"
    >
      <g
        strokeWidth="5"
        transform="translate(2.5 2.5)"
      >
        <circle
          cx="16"
          cy="16"
          r="16"
          strokeOpacity=".5"
        />
        <path
          d="M32 16c0-9.94-8.06-16-16-16"
        >
          <animateTransform
            attributeName="transform"
            dur="1s"
            from="0 16 16"
            repeatCount="indefinite"
            to="360 16 16"
            type="rotate"
          />
        </path>
      </g>
    </g>
  </svg>
</div>
`;
